boefjes/boefjes/plugins/kat_external_db/normalize.py

```diff
@@ -0,0 +1,71 @@
+import json
+import logging
+from ipaddress import IPv4Interface, ip_interface
+from typing import Iterator, Union
+
+from boefjes.job_models import NormalizerMeta
+from octopoes.models import OOI
+from octopoes.models.ooi.dns.zone import Hostname
+from octopoes.models.ooi.network import IPAddressV4, IPAddressV6, IPV4NetBlock, IPV6NetBlock, Network
+
+# Expects raw to be json containing a list of ip_addresses/netblocks
+# (as dictionaries) and a list of domains (as dictionaries).
+# The paths through the dictionaries (to the lists and through the lists)
+# are defined below.
+# T O D O add these variables as normalizer settings in UI.
+IP_ADDRESS_LIST_PATH = ["ip_addresses"]
+IP_ADDRESS_ITEM_PATH = ["ip_address"]
+DOMAIN_LIST_PATH = ["domains"]
+DOMAIN_ITEM_PATH = ["domain"]
+
+
+def follow_path_in_dict(path, path_dict):
+    """Follows a list of keys in a dictionary recursively."""
+    if path:
+        key = path[0]
+        if key not in path_dict:
+            raise KeyError(f"Key {key} not in {list(path_dict.keys())}")
+        return follow_path_in_dict(path=path[1:], path_dict=path_dict[key])
+    return path_dict
+
+
+def run(normalizer_meta: NormalizerMeta, raw: Union[bytes, str]) -> Iterator[OOI]:
+    """Yields hostnames, IPv4/6 addresses or netblocks."""
+    results = json.loads(raw)
+    network = Network(name=normalizer_meta.raw_data.boefje_meta.arguments["input"]["name"])
+    y_addresses, y_blocks, y_hostnames = 0, 0, 0
+
+    for address_item in follow_path_in_dict(path=IP_ADDRESS_LIST_PATH, path_dict=results):
+        address = follow_path_in_dict(path=IP_ADDRESS_ITEM_PATH, path_dict=address_item)
+
+        mask = None
+        if "/" in address:
+            address, mask = address.split("/")
+
+        # Decide whether we yield IPv4 or IPv6.
+        if isinstance(ip_interface(address), IPv4Interface):
+            address_type = IPAddressV4
+            block_type = IPV4NetBlock
+        else:
+            address_type = IPAddressV6
+            block_type = IPV6NetBlock
+
+        ip_address = address_type(address=address, network=network.reference)
+        yield ip_address
+        y_addresses += 1
```

Comment:
I totally agree. However, when adding a NetBlock through the interface, KAT requires a related IPaddress object as start IP. 

![image](https://github.com/minvws/nl-kat-coordination/assets/115991818/9393d3fc-ce44-41a1-a021-8580bdab33a6)

I can remove the yield but I expect this to cause problems due to the current design of the NetBlock

Reply:
> We should not create an address object for masks. If the database has a netblock, we should just create a netblock object and not also just a single IP address from that netblock. And usually this is the network address so it wouldn't make sense to add that as IP address object to OpenKAT. For example with 192.0.2.0/24 we would add the network address 192.0.2.0 which can't be assigned to a host.

192.0.2.0 is a valid ip-address for a /24, however, it cannot be routed to a host, and its a network-identifier. Adding it as such would be cleaner, but maybe we can make the ip-address object smarter (if there is a netblock associated), and figure out it the address is either a network-identifier, or a broadcast address (as .255 would be in this example)?

Reply:
It's technically an IPv4 address, but for the purpose of OpenKAT it is not useful to have that as an IPv4 address object in the database. We can't run boefjes on it. In IPv6 the all-zero address is defined as the subnet-router anycast address (RFC 4291 2.6.1). Adding this address is only makes sense if we would be scanning on the local network as far as I can see.

In my opinion it would be better to define the IPv4 address object as an IPv4 unicast address that can be assigned to a host so that we don't have to exclude/special-case the non-unicast address everywhere. It's outside the scope of this ticket however, so I am resolvcing this conversation.